Frameworks/CoreGraphics/DWriteWrapper.mm

```diff
@@ -96,61 +113,73 @@ bool operator()(const woc::unique_cf<CFStringRef>& item1, const woc::unique_cf<C
+                woc::unique_cf<CFStringRef> uppercaseNameKey(__CFStringCreateUppercaseCopy(info->displayName.get(), locale.get()));
+                ret.emplace(std::move(uppercaseNameKey), info);
+            }
 
-                // For each font in that family, fill out a _DWriteFontProperties
-                auto info = std::make_shared<_DWriteFontProperties>();
-                ComPtr<IDWriteLocalizedStrings> displayName;
-                ComPtr<IDWriteLocalizedStrings> postScriptName;
-                BOOL exist;
+            if (SUCCEEDED(font->GetInformationalStrings(DWRITE_INFORMATIONAL_STRING_POSTSCRIPT_NAME, &postScriptName, &exist)) && exist) {
+                info->postScriptName.reset(static_cast<CFStringRef>(CFRetain(_CFStringFromLocalizedString(postScriptName.Get()))));
+                woc::unique_cf<CFStringRef> uppercaseNameKey(__CFStringCreateUppercaseCopy(info->postScriptName.get(), locale.get()));
+                ret.emplace(std::move(uppercaseNameKey), info);
+            }
 
-                if (SUCCEEDED(font->GetInformationalStrings(DWRITE_INFORMATIONAL_STRING_FULL_NAME, &displayName, &exist)) && exist) {
-                    info->displayName.reset(__CFStringCreateUppercaseCopy(_CFStringFromLocalizedString(displayName.Get()), locale.get()));
-                    woc::unique_cf<CFStringRef> uppercaseNameKey(__CFStringCreateUppercaseCopy(info->displayName.get(), locale.get()));
-                    ret.emplace(std::move(uppercaseNameKey), info);
-                }
+            info->familyName.reset(CFStringCreateCopy(nullptr, familyName));
 
-                if (SUCCEEDED(font->GetInformationalStrings(DWRITE_INFORMATIONAL_STRING_POSTSCRIPT_NAME, &postScriptName, &exist)) &&
-                    exist) {
-                    info->postScriptName.reset(static_cast<CFStringRef>(CFRetain(_CFStringFromLocalizedString(postScriptName.Get()))));
-                    woc::unique_cf<CFStringRef> uppercaseNameKey(__CFStringCreateUppercaseCopy(info->postScriptName.get(), locale.get()));
-                    ret.emplace(std::move(uppercaseNameKey), info);
-                }
+            info->weight = font->GetWeight();
+            info->stretch = font->GetStretch();
+            info->style = font->GetStyle();
+        }
+    }
 
-                info->familyName.reset(CFStringCreateCopy(nullptr, familyName.get()));
+    return ret;
+}
 
-                info->weight = font->GetWeight();
-                info->stretch = font->GetStretch();
-                info->style = font->GetStyle();
-            }
-        }
+/**
+ * Private static map, that maps uppercase display name and postscript name, to a _DWriteFontProperties struct for the corresponding system
+ * font.
+ */
+static const __DWritePropertiesMap c_systemFontPropertiesMap = []() {
+    woc::unique_cf<CFArrayRef> familyNames(_DWriteCopyFontFamilyNames());
+    return __CreatePropertiesMapForFamilyNames(familyNames.get());
```

Comment:
Yes.  This can be a static inside a method and that wont be invoked during dll load time.  There are very limited things you can do in static initialization due to the loader locks. [https://msdn.microsoft.com/en-us/library/windows/desktop/dn633971(v=vs.85).aspx#general_best_practices](https://msdn.microsoft.com/en-us/library/windows/desktop/dn633971(v=vs.85).aspx#general_best_practices)

---
In reply to: [89407585](https://github.com/Microsoft/WinObjC/pull/1454#discussion_r89407585) [](ancestors = 89407585)